storage2/src/state/transaction.rs

```diff
@@ -1,43 +1,33 @@
-use std::collections::HashMap;
-
-use anyhow::Result;
-
-use jmt::storage::{TreeReader, TreeWriter};
-
-use super::{State, StateWrite};
+use super::StateWrite;
 
 /// Represents a transactional set of changes to a `State` fork,
 /// implemented as a RYW cache over a `State`.
-pub(crate) struct Transaction<'a, R: TreeReader + TreeWriter> {
-    // TODO: should higher-level types be used here instead of jmt types?
-    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
-    unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
-    state: &'a State<'a, R>,
+pub struct Transaction {
+    // TODO determine which fields to include (#1490)
+    // cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    // unwritten_changes: Vec<(jmt::KeyHash, jmt::OwnedValue)>,
+    // state: &'a mut State<'a>,
```

Comment:
One way we could enforce that only one `Transaction` is live for each `State` instance at a time would be to add a `&mut State` reference to the `Transaction`, so that it becomes
```
pub struct Transaction<'s> {
    state: &'s mut State,
    // ...
}
```

Question: do we want to do this?

The advantage is that we can guarantee that there's only one `Transaction` at a time, and also that in `Transaction::commit`, we can directly edit the `State` (by having `pub(super) State::foo(&mut self, ...)` methods and calling them in `Transaction::commit`).

The downside is that now, if we pass around a `Transaction`, we're also passing around a lifetime bound. This might be inconvenient if we wanted to, e.g., spawn multiple tasks using the same `Transaction`, but we probably don't want to do that, because of concurrency reasons.

On balance, it seems like this could be a good API choice?

Reply:
I am curious what the general usage pattern in our component code will look like.

i.e. are we generally going to create and commit a transaction within the same method, or would a component ever create a transaction in `begin_block` and only commit it during `end_block`? That would make using the `&mut State` reference difficult.

Reply:
The overall goal is to be able to combine verification and execution, by starting to execute a (blockchain) transaction, but discard any writes it made if it errors during the execution.  The (storage) transaction is a way to accomplish this, by starting a new (storage) transaction when beginning execution of a (blockchain) transaction, then committing it only when the execution completes successfully.

So I don't think we'll ever want to have a component create a `Transaction` itself, since it's only going to be one part of the whole execution.  We'll probably want to have the `App` construct a `Transaction` when starting to process a (blockchain) transaction, then pass it in to all the components.  For `begin_block` and `end_block`, the fallibility / "transactionality" is less important, because that functionality should be infallible anyways, but I think it would make sense to have the `App` construct a `Transaction` at the start of `begin_block`/`end_block` and commit it at the end of the same method.

Reply:
If the `App` owns the `Transaction`, then I think it makes sense to enforce that only one can exist at a time. The fact that it dissuades us from using transactions concurrently sounds like a positive thing IMO.

Reply:
Cool, sounds good to me.